Math/NumberTheory/QuadraticIntegers.hs

```diff
@@ -0,0 +1,191 @@
+{-# LANGUAGE BangPatterns   #-}
+{-# LANGUAGE DataKinds      #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE PolyKinds #-}
+
+module Math.NumberTheory.QuadraticIntegers (
+  QuadraticInteger(..),
+  Sign(..),
+  discriminant,
+  conjugate,
+  norm,
+  trace,
+  quotRemQ,
+  quotQ,
+  remQ,
+  divModQ,
+  divQ,
+  modQ,
+) where
+
+import GHC.TypeNats.Compat
+import Data.Proxy
+
+infix 5 :+
+
+--------------------------------------------------------------
+-- Type level representation of sign
+  
+data Sign = Pos | Neg
+
+class KnownSign (x :: Sign) where
+  signVal :: proxy x -> Sign
+
+instance KnownSign 'Pos where
+  signVal _ = Pos
+
+instance KnownSign 'Neg where
+  signVal _ = Neg
+
+-- | Takes a quadratic integer and returns its sign as an integer
+signToInt :: forall s d . KnownSign s => QuadraticInteger s d -> Integer
+signToInt _ =
+  case signVal (Proxy :: Proxy s) of
+    Pos -> 1
+    Neg -> -1
+  
+---------------------------------------------------------------
+-- Data definitions for quadratic integers
+
+data QuadraticInteger (sign :: Sign) (d :: Nat) =
+  (:+) { rationalComp :: !Integer, irrationalComp :: !Integer } deriving (Eq)
+
+
+{- If the discriminant of a field is equivalent to 1 mod 4
+   then the basis for the ring of integers is strictly
+   larger than a + bx. Need to represent this other basis -}
+
+data BasisType = Integral | Fractional
+
+
+-- | Takes a quadratic int and returns its basis type
+basisType :: forall s d . (KnownNat d, KnownSign s)
+          => QuadraticInteger s d
+          -> BasisType
+basisType z
+  | disc `mod` 4 == 1 = Fractional
+  | otherwise         = Integral
+  where disc = discriminant z
+
+
+instance Show (QuadraticInteger s d) where
+    show (a :+ b)
+        | b == 0     = show a
+        | a == 0     = s ++ b'
+        | otherwise  = show a ++ op ++ b'
+        where
+            b' = if abs b == 1 then "x" else show (abs b) ++ "*x"
+            op = if b > 0 then "+" else "-"
+            s  = if b > 0 then "" else "-"
+
+instance forall s d . (KnownNat d, KnownSign s) => Num (QuadraticInteger s d) where
+  (+) (a :+ b) (c :+ d) = (a+c) :+ (b+d)
+
+  (*) x@(a :+ b) (c :+ d) =
+    case basisType x of
```

Comment:
I am a bit worried that dispatching by `basisType` will take place in runtime instead of compile time. We may introduce a type family like one below, but it will probably fail to compile in older GHCs. 

```haskell
type family GetBasisType (sign :: Sign) (d :: Nat) :: BasisType where
  GetBasisType 'Pos 0 = TypeError ('Text "Discriminant should not be divisible by 4")
  GetBasisType 'Pos 1 = 'Fractional
  GetBasisType 'Pos 2 = 'Integral
  GetBasisType 'Pos 3 = 'Integral
  GetBasisType 'Pos n = GetBasisType 'Pos (n - 4)
  GetBasisType 'Neg 0 = TypeError ('Text "Discriminant should not be divisible by 4")
  GetBasisType 'Neg 1 = 'Integral
  GetBasisType 'Neg 2 = 'Integral
  GetBasisType 'Neg 3 = 'Fractional
  GetBasisType 'Neg n = GetBasisType 'Neg (n - 4)
```